### PR TITLE
fix: make loadDeclarativeJobs idempotent on cron changes and YAML removals

### LINKED ISSUE #640: perf/infra: make loadDeclarativeJobs idempotent on cron changes and YAML removals
## Problem

`upsertDeclarativeJob` uses a unique constraint on `(agent_id, cron_expr, task_payload::text) WHERE created_by = 'system'`. This means:

1. **Cron expression change** — updating the `cron:` field in an agent's YAML creates a *new* `scheduled_jobs` row rather than updating the existing one. The old row stays in `pending` status and keeps firing alongside the new one.

2. **Schedule entry removed from YAML** — removing a `schedule:` entry entirely leaves the old `scheduled_jobs` row permanently in `pending`, firing forever. There is no cleanup step in `loadDeclarativeJobs`.

Both cases were discovered during the T2125 dark-period change (#623), where changing `*/30 * * * *` to `*/30 6-17 * * *` required a one-off migration (042) to cancel the stale row. That migration is a workaround; the platform should handle this transparently.

## Proposed fix

After all `upsertDeclarativeJob` calls complete in `loadDeclarativeJobs`, perform a cleanup pass:

1. Collect the full set of `(agent_id, cron_expr, task_payload::text)` tuples from the current YAML across all agents.
2. Query `scheduled_jobs` for all `created_by = 'system'` rows in `pending` or `failed` status whose `(agent_id, cron_expr, task_payload::text)` triple is **not** in that set — these are stale (changed or removed).
3. Cancel/suspend all stale rows (set `status = 'cancelled'`).

This makes `loadDeclarativeJobs` fully idempotent on restart: cron changes and YAML removals are both handled automatically without migrations or manual DB surgery.

### Edge cases to consider

- A schedule entry whose `agent_id` override points to a *different* agent than the YAML file's `name` — the cleanup must account for `targetAgentId` (as used in `upsertDeclarativeJob`), not just `config.name`.
- Jobs currently in `running` status should not be cancelled — the watchdog handles those via timeout recovery.
- Jobs in `suspended` status (manually paused) should probably also be left alone — the operator may have suspended them intentionally.

## Acceptance criteria

- [ ] On restart, if a schedule entry's `cron` field changed, the old `scheduled_jobs` row is automatically cancelled and only the new row fires
- [ ] On restart, if a `schedule:` entry was removed from a YAML file, the corresponding `scheduled_jobs` row is automatically cancelled
- [ ] Jobs in `running` or `suspended` status are not touched by the cleanup
- [ ] The cleanup is logged at `info` level (agent_id, old cron, action: cancelled) for observability
- [ ] Fresh-install behaviour is unchanged (cleanup finds nothing to cancel, zero rows affected)
- [ ] Existing tests pass; a new integration test covers the stale-row cleanup scenario

### PR DESCRIPTION
**Pull request by josephfung**
## Summary

- Adds `cancelStaleDeclarativeJobs()` to `SchedulerService` — cancels `created_by = 'system'` rows in `pending/failed` status whose `(agent_id, cron_expr, task_payload)` triple is no longer in the live YAML config
- Wires the cleanup into `loadDeclarativeJobs` after all upserts complete, using `::jsonb::text` casting so PostgreSQL normalizes both sides of the exclusion comparison identically
- Covers both failure modes from #640: cron expression changes (new row conflicts on unique index, old row stays pending) and removed schedule entries (no corresponding YAML at all)
- Jobs in `running` or `suspended` status are never touched; `created_by != 'system'` rows (conversation-created, API-created) are never affected

Closes #640

## Test Plan

- [ ] 11 new unit tests added (4 in `scheduler-service.test.ts`, 7 in `scheduler.test.ts`); all 127 scheduler tests pass
- [ ] Edge cases covered: failed upserts excluded from live set, cross-agent `targetAgentId` used (not `config.name`), empty live set calls cleanup with `[]`, cleanup failure logs and does not abort startup
- [ ] Fresh-install behavior unchanged (cleanup finds nothing to cancel, count = 0, no log noise)

___

## **CodeAnt-AI Description**
Prevent stale scheduled jobs from staying active after YAML changes

### What Changed
- After loading schedules from YAML, the app now removes old system-created jobs that no longer match any current schedule
- This covers both changed cron timing and schedules removed from YAML entirely
- Only pending and failed jobs are cancelled; running and suspended jobs are left alone
- If cleanup fails, startup continues and the error is logged instead of stopping the scheduler

### Impact
`✅ Fewer duplicate scheduled jobs after config changes`
`✅ Fewer leftover jobs after removing a schedule`
`✅ Safer scheduler startup when cleanup hits an error`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

## Release Notes

* **Fixed**
  * Declarative jobs now automatically cancel stale scheduled jobs when cron expressions change or when schedule entries are removed from YAML, eliminating the need for manual migrations.

[Review Change Stack](https://app.coderabbit.ai/change-stack/josephfung/curia/pull/656?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by josephfung** [comment]
@coderabbitai review

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,6 +24,7 @@ bus event types) are noted explicitly even in the `0.x` range.
 
 ### Fixed
 
+- **Declarative job cleanup** — `loadDeclarativeJobs` now automatically cancels stale `scheduled_jobs` rows when a cron expression changes or a schedule entry is removed from YAML, eliminating the need for one-off migrations. (#640)
```
[the remark this answers is not in the document]
Comment (josephfung) [reply CHANGELOG.md]: Fixed — shortened to: "auto-cancels stale YAML schedules after cron changes or removals. (#640)"

File: src/scheduler/scheduler-service.ts [hunk 2]
```diff
@@ -549,6 +549,76 @@ export class SchedulerService {
...
+        'Stale declarative job cancelled — schedule entry changed or removed from YAML',
+      );
+    }
+
+    // TODO: if declarative schedules ever use intent_anchor, cancelling the scheduled_jobs row
+    // here leaves the linked agent_tasks row orphaned (scheduled_job_id becomes NULL). At that
+    // point this method should also cancel any agent_tasks rows whose scheduled_job_id was
+    // just cleared — mirror the cleanup in cancelJob(). No declarative schedules use
+    // intent_anchor today, so this is safe to defer.
+    return rows.length;
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/scheduler/scheduler-service.ts:619]: Acknowledged and deferred. No declarative YAML schedules currently use `intent_anchor`, so no `scheduled_jobs` rows created by `upsertDeclarativeJob` have a linked `agent_tasks` row. The TODO comment at the end of `cancelStaleDeclarativeJobs` documents the gap and will be addressed in the PR that introduces the first declarative `intent_anchor` schedule.

File: tests/unit/scheduler/scheduler-service.test.ts [hunk 3]
```diff
@@ -799,6 +799,77 @@ describe('SchedulerService', () => {
...
+      });
+
+      const liveTuples = [
+        { agentId: 'coordinator', cronExpr: '*/30 6-17 * * *', taskPayload: '{"task":"old-cron"}' },
+      ];
+
+      const count = await svc.cancelStaleDeclarativeJobs(liveTuples);
+
+      expect(count).toBe(2);
+      const [sql, params] = pool.query.mock.calls[0] as [string, unknown[]];
```
[the remark this answers is not in the document]
Comment (josephfung) [reply tests/unit/scheduler/scheduler-service.test.ts]: Fixed — changed `const [sql, params] = ...` to `const [sql] = ...` where `params` was unused, and `const [_sql, params] = ...` to `const [, params] = ...` where `_sql` was unused. CI should be green now.

File: src/scheduler/scheduler-service.ts [hunk 4]
```diff
@@ -549,6 +549,76 @@ export class SchedulerService {
...
+      : '';
+
+    const sql = `
+      UPDATE scheduled_jobs
+         SET status = 'cancelled'
+       WHERE created_by = 'system'
+         AND status IN ('pending', 'failed')
+         ${exclusionClause}
+      RETURNING id, agent_id, cron_expr, task_payload
+    `;
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/scheduler/scheduler-service.ts:603]: Acknowledged and deferred. No declarative YAML schedules currently use `intent_anchor`, so no `scheduled_jobs` rows created by `upsertDeclarativeJob` have a linked `agent_tasks` row — there's nothing to orphan today. A TODO comment in `cancelStaleDeclarativeJobs` documents the gap and will be addressed in the PR that introduces the first declarative `intent_anchor` schedule.

File: src/scheduler/scheduler.ts [hunk 5]
```diff
@@ -577,6 +587,25 @@ export class Scheduler {
         );
       }
     }
+
+    // Cancel system-created rows whose (agent_id, cron_expr, task_payload) triple is no longer
+    // declared in any agent YAML. This handles both cron expression changes (old row becomes
+    // stale when the new cron conflicts on the unique index) and removed schedule entries
+    // (row has no corresponding YAML declaration at all).
+    try {
+      const cancelledCount = await this.schedulerService.cancelStaleDeclarativeJobs(liveTuples);
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/scheduler/scheduler.ts]: Valid catch — fixed. The tuple push now happens before the `try` block, and `liveTuples` is renamed `declaredTuples` throughout to make the intent explicit: a YAML declaration protects its DB row from stale-job cleanup regardless of whether the upsert succeeds. The updated test confirms that both the failed and successful upsert's tuples appear in the declared set.
